Create initial functional tests using pytest

davehunt opened the pull request:
This is the first step in replacing the CasperJS tests with pytest tests. Once this has landed we'll work our way through the CasperJS tests and either migrate them to this stack or remove them. We'll then do the same with the mcom-tests.

Feedback would be appreciated from @bobsilverberg, @alexgibson, @pmclanahan and @jgmize. The plan is to run these in Web QA's Jenkins instance in addition to the mcom-tests, but I'd like to also get these running in the new Jenkins pipeline, which @glogiotatidis may be able to help with.

alexgibson commented on docs/testing.rst:
```diff
@@ -8,148 +8,176 @@
-tests, can be installed by running the following command from the root directory of the project::
+First follow the :ref:`installation instructions for bedrock<install>`, which
+will install the specific versions of Jasmine/Karma which are needed to run the
+unit tests, and guide you through installing pip and setting up a virtual
```
should this be `installing peep` instead of `installing pip`?

davehunt replied:
peep doesn't need to be installed as it's in bedrock/bin, and peep depends on pip, so that does still need to be installed.

alexgibson replied:
ah ok, thanks :+1:

alexgibson commented on docs/testing.rst:
```diff
@@ -8,148 +8,176 @@
+.. code-block:: bash
 
-    npm test -- tests/functional --domain=https://www-dev.allizom.org --config=tests/config.json
+    $ py.test --driver Firefox --html results.html -n auto tests/functional/test_newsletter.py
```
When I try to run this line I get the following error:

    usage: py.test [options] [file_or_dir] [file_or_dir] [...]
    py.test: error: unrecognized arguments: -n

If I omit `-n auto` the tests seem to run fine

davehunt replied:
You probably don't have pytest-xdist installed. I added it to the requirements file. While it's not strictly required, it allows you to run the tests in multiple processes, and therefore in parallel.

alexgibson replied:
If I type `pip freeze` it lists `pytest-xdist==1.13.1`. I'll try reinstalling and see if that helps

alexgibson replied:
I uninstalled `pytest-xdist` and then reinstalled requirements but still the same error?

alexgibson commented:
Is it possible to specify the output directory for the HTML report? Perhaps it could live under `/tests/functional`, and we could add it to .gitignore?

alexgibson commented:
Looks like Travis is failing because `python manage.py test` is trying to run the Selenium tests as well as the unit tests?

alexgibson commented:
@davehunt - this looks great! Couple of niggles, but tests are running and passing for me locally. Great job on the docs :+1:

alexgibson commented:
I also got a `pytest` warning, but I assume this is ok for now:

    WI1 /Users/agibson/Sites/bedrock/venv/lib/python2.7/site-packages/pytest_selenium/pytest_selenium.py:89 'pytest_runtest_makereport' hook uses deprecated __multicall__ argument

davehunt commented:
Updated based on your comments @alexgibson and hopefully fixed the Travis issue.

davehunt commented:
The pytest warning should be safe to ignore. I suspect pytest-selenium will support the latest and previous versions of pytest, so will probably drop 2.6.x soon, and once the next version after 2.8.x is out we can remove the reference to the deprecated argument.

alexgibson commented on requirements/test_functional.txt:
```diff
@@ -0,0 +1,8 @@
+# sha256: bwGLPxbC1cSQ-u-cot0Pk4LpMPi3gDp1cVGMTDQE78o
+pytest==2.8.1
```
`test_redirects.txt` uses `pytest 2.7.2` - should we be using the same version for consistency?

davehunt replied:
We could potentially even combine these and have a tests.txt, what do you think?

alexgibson commented:
Seems the only way I can get all the dependencies installed to run the functional tests was to use `pip` instead of `peep`. This was using a fresh venv.

    pip install -r requirements/test_functional.txt

As soon as I did this, everything runs fine. I wonder if peep isn't installing all the required dependencies?

pmclanahan commented:
> Seems like peep isn't installing everything that's needed?

peep does indeed not install anything you don't tell it to. you should pip install into a fresh venv then pip freeze to get the full requirements list.

bobsilverberg commented on tests/functional/test_newsletter.py:
```diff
@@ -0,0 +1,38 @@
+    assert not page.privacy_policy_accepted
+
+
+@pytest.mark.nondestructive
```
Is this really nondestructive? I would have thought that signing up for a mailing list was destructive.

davehunt replied:
I agree, what do you think @alexgibson? This would mean the test wil be skipped for sensitive URLs. I'll also add a regex to setup.cfg to indicate sensitive URLs. Would `mozilla\.(com|org)` be suitable?

alexgibson replied:
Yes I think it makes sense to mark this as destructive :+1:

alexgibson replied:
Also maybe worth adding a note to the guidelines about which type of test should generally not be marked as `nondestructive`?

davehunt replied:
Good idea. I'm going to enhance pytest-selenium to accept a sensitive url regex in setup.cfg. At the moment it requires either either a command line argument or environment variable. Supporting setup.cfg means it can easily be configured for a project and stored in the repo.

davehunt replied:
Documentation updated.

bobsilverberg commented on tests/pages/newsletter.py:
```diff
@@ -0,0 +1,87 @@
+        return el.is_selected()
+
+    def select_html_format(self):
+        self.selenium.find_element(*self._html_format_locator).click()
```
For `accept_privacy_policy`, you ensure that the privacy policy is not already selected, but for `select_html_format` (and `select_text_format`) you do not do the same. Is that intentional? Are these meant as toggles which will simply change the state regardless of the current state?

davehunt replied:
These are radio buttons, so additional clicks do not toggle the selected state.

bobsilverberg commented on tests/pages/page.py:
```diff
@@ -0,0 +1,42 @@
+        self.root = root
+
+        if self.root is None and self._root_locator is not None:
+            self.root = self.selenium.find_element(*self._root_locator)
```
So you could end up with `self.root` == `None`, correct? Is that okay? Is it up to the PageRegion developer to either ensure that isn't the case or to make sure they don't use `self.root`? In the past I think we defaulted `self.root` to `selenium` so that `self.root` was always available. I'm not saying that's what we should do, but I just wanted to check your thinking on this.

davehunt replied:
I'm not keen on `self.root` defaulting to `self.selenium`, though perhaps we could default it to the document root (if that's even possible?). I was thinking that root is totally optional, and for page regions that don't specify it, using `region.root` would not make sense. Instead those page regions should use `self.selenium` directly.

bobsilverberg commented:
@davehunt LGTM. I asked a few questions above.

davehunt commented:
@alexgibson I've updated the requirements file for peep, however having all packages pinned means they'll get out of date regularly. I would recommend using something like https://requires.io/github/mozilla/bedrock/requirements/?branch=master to keep on top of them.

alexgibson commented:
> @alexgibson I've updated the requirements file for peep, however having all packages pinned means they'll get out of date regularly. I would recommend using something like https://requires.io/github/mozilla/bedrock/requirements/?branch=master to keep on top of them.

I think there's still a dependency missing when installing with peep (I'm still seeing `py.test: error: unrecognized arguments: -n`).

As for the suggestion to use something like `requires.io`, this sounds good to me. Although I guess this scenario is true for any project that specifically locks down dependencies with something like peep.

As for the requirements file, my only concern is we currently have pytest v2.7.2 in `test_redirects.txt` and `v2.8.2 in `test_functional.txt`. Would it be better for everything to be on a single version? Combining the two requirements files may be a simple solution.

 Any thoughts or strong feeling on the above, @pmclanahan?

alexgibson commented:
As of the latest commit this all installs and runs fine for my using peep in a fresh venv :+1:

schalkneethling commented:
Installation was quick and easy. Running the tests yields the following though:
https://sneethling.pastebin.mozilla.org/8848680

Is this expected? Did I miss a step? @davehunt

davehunt commented:
@schalkneethling I believe the test failed because you do not have the localisations installed. I found this too with a local instance, and @alexgibson pointed me towards some doc for installing the locales.

pmclanahan commented:
@alexgibson in an ideal world yes, we'd have these requirements synced. But for now it really depends on if they'll all be run by the same system. If these are to be run in travis e.g., then I'd say yes we should sync them and put them in the same file even. If not, then it's fine for now to leave them separate.

schalkneethling commented:
I cloned the locales, reran the tests and all is good.

pmclanahan commented:
We just fixed this in master. l10n files are no longer required for passing
tests (at least the django tests).

davehunt commented:
@pmclanahan do you mean you're not running tests that require l10n, or that the tests are running but don't need to pass? This particular test is checking the functionality of changing the language of the home page, so it will need at least one alternative locale to be present to pass.

pmclanahan commented:
Ah yes. My problem is that I was thinking of the other test suite. Please ignore me :)

alexgibson commented:
@davehunt - how about we combine the two testing requirements files into a combined `testing.txt`. Then we can at least ensure they are both using the same pytest version if someone wants to run both suites of tests? Other than that, I think we're all good here :+1:

davehunt commented:
Rebased and squashed. I also added a default locale (en-US) and the ability to override when creating a page object instance. This is because I noticed different locators for the footer newsletter sign up from en-GB. /cc @alexgibson @bobsilverberg @m8ttyB @schalkneethling @pmclanahan @jgmize @glogiotatidis.

alexgibson commented:
@davehunt - looks like travis is failing because some of the hashes don't match the versions in the updated requirements?

davehunt commented:
Yeah, I'm not sure why Travis is getting different hashes to me..?

> On 14 Oct 2015, at 18:41, Alex Gibson <notifications@github.com> wrote:
> 
> @davehunt - looks like travis is failing because some of the hashes don't match the versions in the updated requirements?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bobsilverberg commented:
@davehunt If you are not already familiar with it, you could try `peepin` [1], a tool written by our own beloved @peterbe, which will generate the appropriate hashes for you. I'm not sure if this will resolve the differences between your local install and Travis, but it might be worth trying.

[1] https://pypi.python.org/pypi/peepin

pmclanahan commented:
`peepin` +1. It'll get hashes for all of the packages. The issue is that PyPI can have different files for the same package (e.g. .tar.gz, .zip, .wheel, etc.). These will all have different hashes, and different versions of pip prefer different ones.

davehunt commented:
Okay, the difference was the `--no-wheel` argument that Travis is passing. I've updated the requirements file to support when this argument is present. I had problems using `peepin` with an existing requirements file, which I'll raise separately.

alexgibson commented:
Travis is happy, everything installs fine for me using a fresh venv, tests all pass! r+ :beers:
